signac/contrib/project.py

```diff
@@ -215,8 +228,11 @@ def __setitem__(self, key, value):
 
 def _invalidate_config_cache(project):
     """Invalidate cached properties derived from a project config."""
+    # TODO: Consider using functools.cached_property for these instead. That simplify our code,
+    # although at the expense of creating multiple locks (each cached_property has its own lock
+    # AFAICT from looking at the implementation).
```

Comment:
> for non-mutable properties one can just use @property and @cache in combination

The docs discussing that combination aren't referring to immutable properties, they're referring to classes where `__slots__` is defined so that instances don't have a `__dict__`. That being said, you are partially correct that this combination would give an immutable property, but there are some dealbreaking caveats. Note that `functools.cache` wasn't introduced until 3.9 so we would have to use `functools.lru_cache` instead, but that's not an issue. The problem will be that the cache will be extremely cumbersome to access, and it will be class-wide and cannot be cleared on a per-instance basis.

> The gain in simplicity might not outweigh the potential performance cost of creating multiple locks.

The actual result will be far fewer locks because it will only create one lock per class, not one per instance. In a GIL-free world that would lead to excessive serialization, but we don't live in that world so only having one instance set the property at a time shouldn't be a problem. And for subsequent accesses the same implementation details that unfortunately make `cached_property` writeable (and that require the existence of `__dict__`) will also make it much faster than a normal `property` for typical usage. Here's a benchmark of the different options including @csadorf's proposal above
```
In [40]: class A:
    ...:     def __init__(self):
    ...:         self._z = None
    ...:
    ...:     @cached_property
    ...:     def w(self):
    ...:         return 0
    ...:
    ...:     @property
    ...:     @lru_cache(None)
    ...:     def x(self):
    ...:         return 0
    ...:
    ...:     @property
    ...:     def y(self):
    ...:         if not hasattr(self, '_y'):
    ...:             self._y = 0
    ...:         return self._y
    ...:
    ...:     @property
    ...:     def z(self):
    ...:         if self._z is None:
    ...:             self._z = 0
    ...:         return self._z
    ...:

In [41]: a = A()

In [42]: %timeit a.w
31.1 ns ± 0.268 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [43]: %timeit a.x
67 ns ± 0.648 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [44]: %timeit a.y
109 ns ± 0.976 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [45]: %timeit a.z
97.9 ns ± 0.86 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

As you can see, `cached_property` is the clear winner, followed by the combination of `property` with `lru_cache`. You'll also see constructors get faster because the initialization only happens once for the class (also true for `property`, but our manual caching introduces extra work per instance).

> unless it improves thread safety.
I think the main benefit to thread safety that we get is that `cached_property`s will automatically get locked without us having to think about it. Conversely, there are some methods that need to be mutually locked and this class won't help us there, so we'll still end up needing to maintain per-instance locks for those cases. That is more what I was referring to in my comment about "the expense of multiple locks". I wasn't so much thinking of performance cost as the fact that multiple locks may not be workable for certain mutually interacting properties/methods.

Reply:
> The docs discussing that combination aren't referring to immutable properties, they're referring to classes where `__slots__` is defined so that instances don't have a `__dict__`. That being said, you are partially correct that this combination would give an immutable property, but there are some dealbreaking caveats. Note that `functools.cache` wasn't introduced until 3.9 so we would have to use `functools.lru_cache` instead, but that's not an issue. The problem will be that the cache will be extremely cumbersome to access, and it will be class-wide and cannot be cleared on a per-instance basis.

Thanks for the clarification, I indeed did not read the preceding paragraph careful enough, however – like you said – this would still be the recommended approach as far as I understand; see also the first paragraph:

> Transform a method of a class into a property whose value is computed once and then cached as a normal attribute for the life of the instance. Similar to [property()](https://docs.python.org/3/library/functions.html#property), with the addition of caching.

 Looks like that approach would already yield a slight performance boost.

Reply:
I'm a little confused about what "this" and "that approach" are referring to in your last comment :joy: 

To make sure that we're on the same page: Using `cached_property` is fine and the recommended approach, and as I showed in my last comment it does lead to a performance boost. Using `property` with `cache` also gives a performance boost. However, I think the fact that it creates a per-class rather than per-instance cache makes it unworkable for our purposes.

Reply:
> Using `cached_property` is fine and the recommended approach

...except that the resulting attribute from `cached_property` is settable, which we don't want. I mentioned this in my first comment and I'm confused about why we're evaluating `cached_property` as a viable alternative to the status quo.

```python
>>> from functools import cached_property
>>> class A:
...   @cached_property
...   def id(self):
...     return "abcdef"
...
>>> a = A()
>>> a.id
'abcdef'
>>> a.id = "this is settable"
>>> a.id
'this is settable'
```

Reply:
I'm a bit confused as well. I don't think that `cached_property` should be discussed as a viable alternative. And why are you saying that `@propert+@cache` is caching class-wide? Here is a simple test which demonstrates that it is per-instance:
```python
from functools import cache, lru_cache


class A:
    @property
    @cache
    def foo(self):
        print("calc foo.", id(self))
        return "bar"
```
When I evaluate this for
```python


a = A()
b = A()

print(a.foo)
print(a.foo)
print(b.foo)
print(b.foo)
```
I get:
```console
% python test.py
calc foo. 140312263293920
bar
bar
calc foo. 140312263292816
bar
bar
```

Reply:
Let me clarify: _I no longer think this comment is actionable in any way, and I will remove it_.

> Using cached_property is fine and the recommended approach

That comment was about what the docs recommend for caching properties, not what I was recommending for signac. I know `cached_property` is settable, but I wasn't thinking of that as a serious problem until @bdice raised it in his first comment. Not every library views it this way (scikit-learn is a prominent example of using raw attributes, for instance, and PyTorch does so in some cases as well), so it doesn't have to be a showstopper. I'm fine with deciding against that approach in signac, though.

> And why are you saying that @propert+@cache is caching class-wide?

The _values_ are per-instance, but the _cache_ is per class. A property is functionally like a `classmethod` that is called on instances. So in your example `a.foo` is really closer to `A.foo(a)`. Therefore, the cache is stored in the class `A` and keyed on the instance of `a` that it is called on. You cannot clear the cache for just a single instance, you have to clear the cache for all instances (and doing so is extremely cumbersome since you have to bypass normal attribute access patterns to touch the cache), and that's a non-starter when for example you have many jobs and you modify one of their statepoints: you don't want to lose cached information on _all_ job instances just because one statepoint was modified.

Reply:
> Not every library views it this way (scikit-learn is a prominent example of using raw attributes, for instance, and PyTorch does so in some cases as well), so it doesn't have to be a showstopper. I'm fine with deciding against that approach in signac, though.

Might not be a showstopper, but since signac has been using immutable properties until now, would be unexpected to change the API in this way, no?

> > And why are you saying that @Propert+@cache is caching class-wide?
> 
> The _values_ are per-instance, but the _cache_ is per class. A property is functionally like a `classmethod` that is called on instances. So in your example `a.foo` is really closer to `A.foo(a)`. Therefore, the cache is stored in the class `A` and keyed on the instance of `a` that it is called on. You cannot clear the cache for just a single instance, you have to clear the cache for all instances (and doing so is extremely cumbersome since you have to bypass normal attribute access patterns to touch the cache), and that's a non-starter when for example you have many jobs and you modify one of their statepoints: you don't want to lose cached information on _all_ job instances just because one statepoint was modified.

I get what you mean now. That said, I do wonder though whether the performance penalty might be acceptable, since modifying the state point shouldn't be a routine operation, should it?

Reply:
> Might not be a showstopper, but since signac has been using immutable properties until now, would be unexpected to change the API in this way, no?

Honestly I'm not sure that's true. You can't effectively query `help` on properties or attributes since you'll just get the docstring for the returned value, so unless someone goes to our online documentation and actively searches for the specific docs they won't even notice the change. It will largely only be relevant for new users, who won't have seen the original code. I admit that it is potentially a way for users to do something nonsensical, but again I think plenty of libraries get away with this and the potential performance benefits are compelling.

> I get what you mean now. That said, I do wonder though whether the performance penalty might be acceptable, since modifying the state point shouldn't be a routine operation, should it?

That is a good point. The tradeoff is probably worth it. Part of me wonders if any of the caching questions here are even relevant for the benchmarks @bdice has looked at most frequently since IIRC those almost always involve a single pass (or maybe a very small number of passes) through the data space, in which case the caching may not be a huge performance improvement either. We need to evaluate concrete benchmarks to make a call here.

Reply:
In the interest of moving forward, I'm going to resolve this and merge the PR. We can benchmark alternatives at a later point.